3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/pii/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,301 @@
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+package pii
+
+// Scala
+import scala.collection.JavaConverters._
+import scala.collection.mutable.MutableList
+
+// Scala libraries
+import org.json4s
+import org.json4s.{DefaultFormats, JValue}
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.jackson.Serialization.write
+
+// Java
+import org.apache.commons.codec.digest.DigestUtils
+
+// Java libraries
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ArrayNode, ObjectNode, TextNode}
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath}
+import com.jayway.jsonpath.MapFunction
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+// Iglu
+import iglu.client.validation.ProcessingMessageMethods._
+import iglu.client.{SchemaCriterion, SchemaKey}
+
+// This project
+import common.ValidatedNelMessage
+import common.utils.ScalazJson4sUtils
+import common.outputs.EnrichedEvent
+
+/**
+ * Companion object. Lets us create a PiiPseudonymizerEnrichment
+ * from a JValue.
+ */
+object PiiPseudonymizerEnrichment extends ParseableEnrichment {
+
+  implicit val json4sFormats = DefaultFormats
+
+  override val supportedSchema =
+    SchemaCriterion("com.snowplowanalytics.snowplow.enrichments", "pii_enrichment_config", "jsonschema", 2, 0, 0)
+
+  def parse(config: JValue, schemaKey: SchemaKey): ValidatedNelMessage[PiiPseudonymizerEnrichment] = {
+    for {
+      conf <- matchesSchema(config, schemaKey)
+      enabled = ScalazJson4sUtils.extract[Boolean](conf, "enabled").toOption.getOrElse(false)
+      emitIdentificationEvent = ScalazJson4sUtils
+        .extract[Boolean](conf, "emitIdentificationEvent")
+        .toOption
+        .getOrElse(false)
+      piiFields        <- ScalazJson4sUtils.extract[List[JObject]](conf, "parameters", "pii").leftMap(_.getMessage)
+      hashFunctionName <- extractStrategyFunction(config)
+      hashFunction     <- getHashFunction(hashFunctionName)
+      piiFieldList     <- extractFields(piiFields)
+    } yield
+      if (enabled)
+        PiiPseudonymizerEnrichment(piiFieldList,
+                                   emitIdentificationEvent,
+                                   PiiStrategyPseudonymize(hashFunctionName, hashFunction))
+      else
+        PiiPseudonymizerEnrichment(List(),
+                                   emitIdentificationEvent = false,
+                                   PiiStrategyPseudonymize(hashFunctionName, hashFunction))
+  }.leftMap(_.toProcessingMessageNel)
+
+  private[pii] def getHashFunction(strategyFunction: String): Validation[String, DigestFunction] =
+    strategyFunction match {
+      case "MD2" => { (b: Array[Byte]) =>
```

Comment:
@chuwy This doesn't seem to work:
```
iiPseudonymizerEnrichment.scala:88: ambiguous reference to overloaded definition,
[error] both method md2Hex in object DigestUtils of type (x$1: String)String
[error] and  method md2Hex in object DigestUtils of type (x$1: java.io.InputStream)String
[error] match expected type ?
[error]       case "MD2"     => { DigestUtils.md2Hex }.success
```

Reply:
Yep, seems `.success` is the problem. You can use `case "MD2" => { DigestUtils.md2Hex _ }.success` or leave as is.

Reply:
I don't think that was the problem:
```
PiiPseudonymizerEnrichment.scala:82: ambiguous reference to overloaded definition,
[error] both method md2Hex in object DigestUtils of type (x$1: String)String
[error] and  method md2Hex in object DigestUtils of type (x$1: java.io.InputStream)String
[error] match expected type ?
[error]       case "MD2"     => { DigestUtils.md2Hex _ }.success
```
The errors seem to suggest that because `md2Hex` is polymorphic the type of the argument is required.

Reply:
Ah, ok. So final `case "MD2"     => { DigestUtils.md2Hex(_: String) }.success` won't be much better than original solution. Up to you.

Reply:
`case "MD2" => { DigestUtils.md2Hex(_: Array[Byte]) }.success` is better. Thanks